### DISCUSSION TITLE
How Do I Get External Libraries to Work?

### DISCUSSION THREAD

**Discussion by sofalights** [pybricks/support#1914] [Q&A] [answered]
I've been trying to get numpy and scikit-learn to work, since I'm using them to control a line-following robot. However, the following error occurs:
```
Traceback (most recent call last):
  File "betterRobo_wrgLRwehe.py", line 1, in <module>
ImportError: no module named 'numpy'
```
Does anyone know how to get them working? Thanks!
Code in question:
```
import numpy as np
from sklearn.linear_model import LogisticRegression
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor, ColorSensor, UltrasonicSensor
from pybricks.parameters import Port
from pybricks.tools import wait, StopWatch
import time
```

**Comment by laurensvalk** [comment]
Pybricks uses MicroPython, which is a small implementation of Python that can run on microcontrollers like those found in SPIKE Prime. Full Python with libraries like `numpy` require an actual computer like your PC or laptop (and can't run on the robot).

That said, line-following can work perfectly find without `numpy`. It all depends on what you want to do :)

**Reply by sofalights** [reply]
I've been using numpy to put the arrays for the logistic regression model (using scikit). 
Here's the code lines where I use numpy:
```
# array elements are [H_left, S_left, V_left, H_right, S_right, V_right]
sensor_data = np.array([
    [0, 0, 0, 0, 0, 0],        
    [140, 100, 100, 0, 0, 0],  
    [0, 0, 0, 140, 100, 100],  
    [140, 100, 100, 140, 100, 100],  
    [250, 10, 90, 250, 10, 90],      
    [120, 10, 100, 120, 10, 100],    
    [0, 0, 0, 180, 100, 100],  
    [180, 100, 100, 0, 0, 0],  
])

# corresponding actions based on array (assign action val to array)
actions = np.array([5, 0, 1, 4, 5, 8, 7, 6])  #action label(in order of array appearance)
```

The elements are placeholders for now
I don't know if it is possible to use tuples though

**Reply by laurensvalk** [reply, marked as answer]
Yes, you can make lists or tuples, and even a Matrix. You can do basic Matrix operations like addition or multiplication but not inversion/regression etc.

In your case, if you just want to store values like that, you don't need the arrays. You can just make lists of lists:

```
sensor_data =[
    [0, 0, 0, 0, 0, 0],        
    [140, 100, 100, 0, 0, 0],  
    [0, 0, 0, 140, 100, 100],  
    [140, 100, 100, 140, 100, 100],  
    [250, 10, 90, 250, 10, 90],      
    [120, 10, 100, 120, 10, 100],    
    [0, 0, 0, 180, 100, 100],  
    [180, 100, 100, 0, 0, 0],  
]
```
